src/NuGet.Clients/PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -495,26 +478,67 @@ protected static IReadOnlyList<PreviewResult> GetPreviewResults(IEnumerable<Reso
+        private async Task<List<IPackageSearchMetadata>> GetPackageMetadataAsync(
+            INuGetUI uiService, 
+            IEnumerable<PackageIdentity> packages, 
+            CancellationToken token)
         {
-            var sources = _sourceProvider.GetRepositories().Where(e => e.PackageSource.IsEnabled);
-
             var results = new List<IPackageSearchMetadata>();
-            foreach (var package in packages)
+
+            // local sources
+            var sources = new List<SourceRepository>();            
+            sources.Add(_packageManager.PackagesFolderSourceRepository);
+            sources.AddRange(_packageManager.GlobalPackageFolderRepositories);
+
+            var allPackages = packages.ToArray();
+
+            // first check all the packages with local sources.
+            var tasks = allPackages.Select(package => GetPackageMetadataAsync(sources, package, token)).ToArray();
+
+            var ignore = tasks
+                .Select(task => task.ContinueWith(newTask => LogError(newTask, uiService), TaskContinuationOptions.OnlyOnFaulted))
+                .ToArray();
+
+            var completed = (await Task.WhenAll(tasks)).Where(m => m != null).ToArray();
+
+            results.AddRange(completed);
+
+            if (completed.Length != allPackages.Length)
             {
-                var metadata = await GetPackageMetadataAsync(sources, package, token);
-                if (metadata == null)
-                {
-                    throw new InvalidOperationException(
-                        string.Format("Unable to find metadata of {0}", package));
-                }
+                // get remaining package's metadata from remote repositories
+                var remainingPackages = allPackages.Where(package => !completed.Any(pack => pack.Identity.Equals(package)));
+
+                tasks = remainingPackages.Select(package => GetPackageMetadataAsync(uiService.ActiveSources, package, token)).ToArray();
+
+                ignore = tasks
+                .Select(task => task.ContinueWith(newTask => LogError(newTask, uiService), TaskContinuationOptions.OnlyOnFaulted))
+                .ToArray();
 
-                results.Add(metadata);
+                var remoteResults = (await Task.WhenAll(tasks)).Where(m => m != null).ToArray();
+
+                results.AddRange(remoteResults);
+            }
+
+            // check if missing metadata for any package
+            if (allPackages.Length != results.Count)
+            {
+                var package = allPackages.First(pkg => !results.Any(result => result.Identity.Equals(pkg)));
+
+                throw new InvalidOperationException(
+                        string.Format("Unable to find metadata of {0}", package));
             }
 
             return results;
         }
 
+        private void LogError(Task task, INuGetUI uiService)
+        {
+            foreach (var ex in task.Exception.Flatten().InnerExceptions)
```

Comment:
There is an Exception Utility class somewhere to help with this. Use that since reading aggregate exceptions can be tricky